src/test/java/com/google/firebase/projectmanagement/FirebaseProjectManagementServiceImplTest.java

```diff
@@ -0,0 +1,911 @@
+/*
+ * Copyright 2018 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.firebase.projectmanagement;
+
+import static com.google.firebase.projectmanagement.FirebaseProjectManagementServiceImpl.FIREBASE_PROJECT_MANAGEMENT_URL;
+import static com.google.firebase.projectmanagement.FirebaseProjectManagementServiceImpl.MAXIMUM_LIST_APPS_PAGE_SIZE;
+import static com.google.firebase.projectmanagement.FirebaseProjectManagementServiceImpl.PATCH_OVERRIDE_KEY;
+import static com.google.firebase.projectmanagement.FirebaseProjectManagementServiceImpl.PATCH_OVERRIDE_VALUE;
+import static com.google.firebase.projectmanagement.ShaCertificateType.SHA_1;
+import static com.google.firebase.projectmanagement.ShaCertificateType.SHA_256;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+
+import com.google.api.client.googleapis.util.Utils;
+import com.google.api.client.http.HttpRequest;
+import com.google.api.client.http.HttpResponse;
+import com.google.api.client.http.HttpResponseInterceptor;
+import com.google.api.client.json.JsonParser;
+import com.google.api.client.testing.http.MockHttpTransport;
+import com.google.api.client.testing.http.MockLowLevelHttpResponse;
+import com.google.api.client.util.Base64;
+import com.google.common.base.Charsets;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.firebase.FirebaseApp;
+import com.google.firebase.FirebaseOptions;
+import com.google.firebase.TestOnlyImplFirebaseTrampolines;
+import com.google.firebase.auth.MockGoogleCredentials;
+import com.google.firebase.testing.MultiRequestMockHttpTransport;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+
```

Comment:
There's a ton of code duplication in this file. Can we do something about it? One simple solution is to combine some tests into a single method as follows (you might have to change how mock responses are set up):

```
  @Test
  public void getIosApp() throws Exception {
    String expectedUrl = String.format(
        "%s/v1beta1/projects/-/iosApps/%s", FIREBASE_PROJECT_MANAGEMENT_URL, IOS_APP_ID);
    firstRpcResponse.setContent(String.format(GET_IOS_RESPONSE, DISPLAY_NAME));
    serviceImpl = initServiceImpl(firstRpcResponse, interceptor);
    IosAppMetadata iosAppMetadata = serviceImpl.getIosApp(IOS_APP_ID);
     checkRequestHeader(expectedUrl, HttpMethod.GET);
     assertEquals(iosAppMetadata, IOS_APP_METADATA);
  
     iosAppMetadata = serviceImpl.getIosAppAsync(IOS_APP_ID).get();
     checkRequestHeader(expectedUrl, HttpMethod.GET);
     assertEquals(iosAppMetadata, IOS_APP_METADATA);
  }
```